Content.Server/Research/Systems/ResearchSystem.cs

```diff
@@ -1,97 +1,74 @@
             InitializeConsole();
-            InitializeServer();
+            InitializeSource();
             InitializeTechnology();
         }
 
-        public bool RegisterServer(ResearchServerComponent server)
-        {
-            if (_servers.Contains(server)) return false;
-            _servers.Add(server);
-            _servers[^1].Id = _servers.Count - 1;
-            return true;
-        }
-
-        public void UnregisterServer(ResearchServerComponent server)
-        {
-            _servers.Remove(server);
-        }
-
         public ResearchServerComponent? GetServerById(int id)
         {
-            foreach (var server in Servers)
+            foreach (var server in EntityQuery<ResearchServerComponent>())
             {
-                if (server.Id == id) return server;
+                if (server.Id == id)
+                    return server;
             }
 
             return null;
         }
 
         public string[] GetServerNames()
         {
-            var list = new string[Servers.Count];
+            var allServers = EntityQuery<ResearchServerComponent>().ToArray();
+            var list = new string[allServers.Length];
 
-            for (var i = 0; i < Servers.Count; i++)
+            for (var i = 0; i < allServers.Length; i++)
             {
-                list[i] = Servers[i].ServerName;
+                list[i] = allServers[i].ServerName;
             }
 
             return list;
         }
 
         public int[] GetServerIds()
         {
-            var list = new int[Servers.Count];
+            var allServers = EntityQuery<ResearchServerComponent>().ToArray();
+            var list = new int[allServers.Length];
 
-            for (var i = 0; i < Servers.Count; i++)
+            for (var i = 0; i < allServers.Length; i++)
             {
-                list[i] = Servers[i].Id;
+                list[i] = allServers[i].Id;
             }
 
             return list;
         }
 
         public override void Update(float frameTime)
         {
-            _timer += frameTime;
-
-            while (_timer > ResearchConsoleUIUpdateTime)
+            foreach (var server in EntityQuery<ResearchServerComponent>())
             {
-                foreach (var server in _servers)
-                {
-                    UpdateServer(server, ResearchConsoleUIUpdateTime);
-                }
-
-                foreach (var console in EntityManager.EntityQuery<ResearchConsoleComponent>())
-                {
-                    if (!_uiSystem.IsUiOpen(console.Owner, ResearchConsoleUiKey.Key)) continue;
-                    UpdateConsoleInterface(console);
-                }
+                if (server.NextUpdateTime > _timing.CurTime)
+                    continue;
+                server.NextUpdateTime = _timing.CurTime + server.ResearchConsoleUpdateTime;
 
-                _timer -= ResearchConsoleUIUpdateTime;
+                UpdateServer(server, (int) server.ResearchConsoleUpdateTime.TotalSeconds);
```

Comment:
Really the best thing to do is derive the points from the last time it was modified, that way the client can just smoothly lerp it on its own and the server sends 0 data, but I assume that's out of scope as that's a bit more work. Cargo also needs updating to use this method.

Reply:
I was informed that the reason it shouldn't be done like this is because things like artifacts will make the point amount suddenly spike for only a single second. I was told this was undesirable behavior.